File: backend/src/main/java/com/university/inventory/service/EquipmentService.java
package com.university.inventory.service;

import com.university.inventory.dto.common.PageResponse;
import com.university.inventory.dto.equipment.EquipmentRequest;
import com.university.inventory.dto.equipment.EquipmentResponse;
import com.university.inventory.entity.Category;
import com.university.inventory.entity.Equipment;
import com.university.inventory.entity.Room;
import com.university.inventory.entity.User;
import com.university.inventory.entity.enums.EquipmentStatus;
import com.university.inventory.exception.ConflictException;
import com.university.inventory.exception.NotFoundException;
import com.university.inventory.mapper.EquipmentMapper;
import com.university.inventory.repository.CategoryRepository;
import com.university.inventory.repository.EquipmentRepository;
import com.university.inventory.repository.RoomRepository;
import com.university.inventory.repository.UserRepository;
import com.university.inventory.service.spec.EquipmentSpecifications;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class EquipmentService {

    private final EquipmentRepository equipmentRepository;
    private final CategoryRepository categoryRepository;
    private final RoomRepository roomRepository;
    private final UserRepository userRepository;
    private final EquipmentMapper equipmentMapper;

    public PageResponse<EquipmentResponse> search(String term, Long categoryId, Long roomId,
                                                  EquipmentStatus status, Pageable pageable) {
        Specification<Equipment> spec = EquipmentSpecifications.combine(
                EquipmentSpecifications.search(term),
                EquipmentSpecifications.hasCategory(categoryId),
                EquipmentSpecifications.hasRoom(roomId),
                EquipmentSpecifications.hasStatus(status));
        Page<EquipmentResponse> page = equipmentRepository.findAll(spec, pageable)
                .map(equipmentMapper::toResponse);
        return PageResponse.of(page);
    }

    public EquipmentResponse findById(Long id) {
        return equipmentMapper.toResponse(get(id));
    }

    @Transactional
    public EquipmentResponse create(EquipmentRequest req) {
        if (equipmentRepository.existsByInventoryNumberIgnoreCase(req.inventoryNumber())) {
            throw new ConflictException("Inventory number already exists: " + req.inventoryNumber());
        }
        Equipment e = new Equipment();
        apply(e, req);
        return equipmentMapper.toResponse(equipmentRepository.save(e));
    }

    @Transactional
    public EquipmentResponse update(Long id, EquipmentRequest req) {
        Equipment e = get(id);
        if (!e.getInventoryNumber().equalsIgnoreCase(req.inventoryNumber())
                && equipmentRepository.existsByInventoryNumberIgnoreCase(req.inventoryNumber())) {
            throw new ConflictException("Inventory number already exists: " + req.inventoryNumber());
        }
        apply(e, req);
        return equipmentMapper.toResponse(e);
    }

    @Transactional
    public void delete(Long id) {
        equipmentRepository.delete(get(id));
    }

    private void apply(Equipment e, EquipmentRequest req) {
        e.setName(req.name());
        e.setInventoryNumber(req.inventoryNumber());
        e.setSerialNumber(req.serialNumber());
        e.setStatus(req.status());
        e.setCondition(req.condition());
        e.setPurchaseDate(req.purchaseDate());
        e.setPrice(req.price());
        e.setDescription(req.description());
        e.setCategory(req.categoryId() == null ? null : categoryRepository.findById(req.categoryId())
                .orElseThrow(() -> NotFoundException.of("Category", req.categoryId())));
        e.setRoom(req.roomId() == null ? null : roomRepository.findById(req.roomId())
                .orElseThrow(() -> NotFoundException.of("Room", req.roomId())));
        e.setResponsibleUser(req.responsibleUserId() == null ? null
                : userRepository.findById(req.responsibleUserId())
                .orElseThrow(() -> NotFoundException.of("User", req.responsibleUserId())));
    }

    private Equipment get(Long id) {
        return equipmentRepository.findById(id)
                .orElseThrow(() -> NotFoundException.of("Equipment", id));
    }
}

File: frontend/src/app/core/services/equipment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { API_URL } from '../config';
import { Equipment, EquipmentRequest, EquipmentStatus, PageResponse } from '../models/models';

export interface EquipmentQuery {
  search?: string;
  categoryId?: number | null;
  roomId?: number | null;
  status?: EquipmentStatus | null;
  page?: number;
  size?: number;
  sort?: string;
}

@Injectable({ providedIn: 'root' })
export class EquipmentService {
  private readonly base = `${API_URL}/equipment`;
  constructor(private http: HttpClient) {}

  list(q: EquipmentQuery): Observable<PageResponse<Equipment>> {
    let p = new HttpParams();
    if (q.search) p = p.set('search', q.search);
    if (q.categoryId) p = p.set('categoryId', q.categoryId);
    if (q.roomId) p = p.set('roomId', q.roomId);
    if (q.status) p = p.set('status', q.status);
    p = p.set('page', q.page ?? 0).set('size', q.size ?? 10);
    if (q.sort) p = p.set('sort', q.sort);
    return this.http.get<PageResponse<Equipment>>(this.base, { params: p });
  }

  get(id: number): Observable<Equipment> {
    return this.http.get<Equipment>(`${this.base}/${id}`);
  }

  create(body: EquipmentRequest): Observable<Equipment> {
    return this.http.post<Equipment>(this.base, body);
  }

  update(id: number, body: EquipmentRequest): Observable<Equipment> {
    return this.http.put<Equipment>(`${this.base}/${id}`, body);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${id}`);
  }
}
